packages/assets-controllers/src/DeFiPositionsController/DeFiPositionsController.ts

```diff
@@ -0,0 +1,165 @@
+  ControllerStateChangeEvent<
+    typeof controllerName,
+    DeFiPositionsControllerState
+  >;
+
+/**
+ * The external actions available to the {@link DeFiPositionsController}.
+ */
+export type AllowedActions =
+  | AccountsControllerGetAccountAction
+  | AccountsControllerGetSelectedAccountAction;
+
+/**
+ * The external events available to the {@link DeFiPositionsController}.
+ */
+export type AllowedEvents =
+  | NetworkControllerStateChangeEvent
+  | AccountsControllerSelectedAccountChangeEvent;
+
+/**
+ * The messenger of the {@link DeFiPositionsController}.
+ */
+export type DeFiPositionsControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  DeFiPositionsControllerActions | AllowedActions,
+  DeFiPositionsControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * Controller that stores assets and exposes convenience methods
+ */
+export class DeFiPositionsController extends BaseController<
+  typeof controllerName,
+  DeFiPositionsControllerState,
+  DeFiPositionsControllerMessenger
+> {
+  // TODO: Confirm whether we can store the account address instead of the id
+  // Storing the address means we don't need to query it in every event handler
+  #selectedAccountId: string;
+
+  /**
+   * Tokens controller options
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - The controller messenger.
+   * @param options.state - Initial state to set on this controller.
+   */
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: DeFiPositionsControllerMessenger;
+    state?: Partial<DeFiPositionsControllerState>;
+  }) {
+    super({
+      name: controllerName,
+      metadata: controllerMetadata,
+      messenger,
+      state: {
+        ...getDefaultDefiPositionsControllerState(),
+        ...state,
+      },
+    });
+
+    this.#selectedAccountId = this.messagingSystem.call(
+      'AccountsController:getSelectedAccount',
+    ).id;
+
+    this.messagingSystem.subscribe(
+      'AccountsController:selectedAccountChange',
+      async (selectedAccount) => {
+        this.#selectedAccountId = selectedAccount.id;
+
+        await this.#updateAccountPositions(selectedAccount.address);
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
+      async () => {
+        const selectedAddress = this.messagingSystem.call(
+          'AccountsController:getAccount',
+          this.#selectedAccountId,
+        );
+
+        if (selectedAddress) {
+          await this.#updateAccountPositions(selectedAddress.address);
+        }
+      },
+    );
+  }
+
+  // TODO: If this becomes an action, accountAddress needs to be inferred from the id
+  async #updateAccountPositions(accountAddress: string) {
+    // TODO: This is done to give the UI a loading effect. Probably not the best way to do this
+    this.update((state) => {
+      state.allDeFiPositions[accountAddress] = null;
+    });
```

Comment:
Is this a good idea? In order to show in the UI that we are currently loading positions for that account, we are wiping them.

Reply:
We use this pattern for notifications too!

This is because the "controllers" on extension are ran in the background, whereas if you were to tie this to the UI, it could be lost/torn down when the extension is minimised or closed.

Just make sure that this "fetching state" is correctly emptied on initialisation.